src/lib.rs

```diff
@@ -64,18 +64,42 @@ impl<T> Arena<T> {
 
     /// Uses the contents of an iterator to allocate values in the arena.
     /// Returns a mutable slice that contains these values.
-    pub fn alloc_extend<I>(&self, iter: I) -> &mut [T]
-        where I: Iterator<Item = T> + ExactSizeIterator
+    pub fn alloc_extend<I>(&self, iterable: I) -> &mut [T]
+        where I: IntoIterator<Item = T>
     {
+        let mut iter = iterable.into_iter();
+
         let mut chunks = self.chunks.borrow_mut();
 
-        if chunks.current.len() + iter.len() > chunks.current.capacity() {
-            chunks.reserve(iter.len());
+        let iter_min_len = iter.size_hint().0;
+        let mut next_item_index;
+        if chunks.current.len() + iter_min_len > chunks.current.capacity() {
+            chunks.reserve(iter_min_len);
+            chunks.current.extend(iter);
+            next_item_index = 0;
+        } else {
+            next_item_index = chunks.current.len();
+            let mut i = 0;
+            while let Some(elem) = iter.next() {
+                if i == iter_min_len {
```

Comment:
I think you could compare `iter_min_len` or `iter_max_len` here, whichever fits in the chunk's remaining space, right?